src/utils/events/eventListeners/eventListenerGenerator.ts

```diff
@@ -32,11 +34,38 @@ export const eventListenerGenerator = async (
     contractAddress ? `filtering Address: ${contractAddress}` : '',
   );
 
-  provider.on(filter, async (log: Log) => {
+  const handleEvent = async (log: Log): Promise<void> => {
     await addEventToQueue(client, clientType, contractAddress, log, provider);
-  });
+  };
+
+  provider.on(filter, handleEvent);
+
+  const listenerRemover = (): void => {
+    verbose(
+      'Removed listener for Event:',
+      eventSignature,
+      contractAddress ? `filtering Address: ${contractAddress}` : '',
+    );
+    provider.off(filter, handleEvent);
+  };
+
+  if (saveRemoverInContext) {
+    const listenerRemoverKey = generateListenerRemoverKey(
+      contractAddress,
+      clientType,
+      eventSignature,
+    );
+    ctx.listenerRemovers[listenerRemoverKey] = listenerRemover;
```

Comment:
I can imagine there could be a `registerListener` method that would accept some parameters, then generate the key and add it to the "cache" outside of this function

Reply:
How could this logic be moved outside of the function? We need a reference to the exact event handler we're removing. Previously I accessed this by returning it from the `eventListenerGenerator` but based on your feedback have moved it inside.

Reply:
What I imagined was something like this:

```js
// listenerRemover.ts 
const removers = {}; // don't need context, this can be internal to the module

export const saveRemover = (contractAddress, eventSignature, listenerRemover) => {
  const key = generateListenerRemoverKey(...)
  removers[key] = listenerRemover;
}
```

Then your code in the generator shortens to declaring the remover and calling the relevant function. There's no need for the `eventListenerGenerator` to know about how the removers are stored/accessed